vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/KMeansPlusPlusClusterInitializer.java

```diff
@@ -0,0 +1,137 @@
+package jetbrains.vectoriadb.index;
+
+import org.apache.commons.rng.simple.RandomSource;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.ValueLayout;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * A smarter than random centroid initialization technique.
+ * 1. Initialize the first centroid randomly.
+ * 2. Calculate the distance to the closest already chosen centroid `distanceToClosestCentroid` for all the vectors.
+ * 3. Choose a vector to become a new centroid based on the weighted probability proportional to `distanceToClosestCentroid^2`.
+ * 4. Repeat from 2 until all the centroid initialized.
+ *
+ * It is possible to adjust the behaviour by setting a random source.
+ * */
+public class KMeansPlusPlusClusterInitializer implements ClusterInitializer {
+    @NotNull
+    private RandomSource randomSource;
+
+    KMeansPlusPlusClusterInitializer() {
+        randomSource = RandomSource.XO_RO_SHI_RO_128_PP;
+    }
+
+    KMeansPlusPlusClusterInitializer(@NotNull RandomSource randomSource) {
+        this.randomSource = randomSource;
+    }
+
+    public void initializeCentroids(
+            MemorySegment pqVectors,
+            long numVectors,
+            float[] distanceTable,
+            int quantizersCount,
+            int codeBaseSize,
+            byte[] pqCentroids,
+            int numClusters,
+            @NotNull ProgressTracker progressTracker
+    ) {
+        progressTracker.pushPhase("Centroids initialization");
+        try (var arena = Arena.openShared()) {
+            var rng = randomSource.create();
+
+            // initialize the first centroid randomly
+            var firstCentroidIndex = rng.nextLong(numVectors);
+            MemorySegment.copy(pqVectors, ValueLayout.JAVA_BYTE, firstCentroidIndex * quantizersCount, pqCentroids, 0, quantizersCount);
+
+            // here we keep distance to the closest centroid for every vector
+            var distancesToClosestCentroid = arena.allocate(numVectors * Float.BYTES, ValueLayout.JAVA_FLOAT.byteAlignment());
+            // we do not know vectors' distances to any centroids, so make them infinite
+            for (long vectorIdx = 0; vectorIdx < numVectors; vectorIdx++) {
+                distancesToClosestCentroid.setAtIndex(ValueLayout.JAVA_FLOAT, vectorIdx, Float.MAX_VALUE);
+            }
+
+            var cores = (int) Math.min(Runtime.getRuntime().availableProcessors(), numVectors);
```

Comment:
Hi Kirill.

Let me outline the implemented algorithm, if I understood it correctly:

1. Choose centroids at random.

In cycle:
2. Choose the n-th centroid.

For all vectors
3. If the distance from the vector to the chosen centroid is the closest, remember it in an array of the closest distances.
4. Calculate the sum of the closest squire distances.

5. Choose a new vector for the n-th centroid using weighted probability over squire distances.


K-Means ++

In a loop:

1. Choose the point that is not a centroid at the moment
2. Find the squire distance to the closest centroid from all centroids.

3. Choose a new vector for the n-th centroid using weighted probability over squire distances.
4. Repeat the loop till not all centroids are chosen

I have a perception that implementation is incorrect because:

1. Check (1) described in the algorithm not implemented.
2. We do not base our choice on the closest distance to **all** centroids because the closest distance is unknown on the n-th step. We must repeat the algorithm n-th times (till the end when all new centroids are already chosen) to know the distance to the closest centroid.

Please prove to me that I am wrong.

Reply:
Hmmm...

> 1. Check (1) described in the algorithm not implemented.

That is a good catch! 

Actually, we do this check implicitly. Here is what I mean:
* When we choose the first centroid randomly, it is obviously not a centroid because there are no centroids yet.
* When we choose all the other centroids, the min distance for all already chosen vectors-centroids will be 0 because we calculate the distance for them with themselves. So when they take part in the random selection, they are ignored because their distances = 0. A distance that is 0 does not affect the counter that is used to find a new centroid.

I see the only option when we can choose a single vector twice to be a centroid. We need the following conditions:
* Random generator gives us the biggest possible value that is = the whole sum
* The last vector in the list is already chosen as a centroid
* The error in the floating-point sum operations gives us a sum that is less than the real sum, so we have to take the list vector blindly.

As we can guess, it is hardly possible. But I agree, it is still possible, so we have to explicitly make sure that will not happen.

I will fix it.

> 2. We do not base our choice on the closest distance to all centroids because the closest distance is unknown on the n-th step. We must repeat the algorithm n-th times (till the end when all new centroids are already chosen) to know the distance to the closest centroid.

We do base our choice on the closest distance to already chosen centroids.

For this purpose, we use `distancesToClosestCentroid` array. Initially, all the distances `infinity` because we do not have any centroids. But on every step, we calculate the distance from every vector to the centroid chosen in the previous step and update `distancesToClosestCentroid` array if it is necessary (if the distance becomes smaller).

So here, it seems ok to me.

Reply:
Checking that we do not make a single vector centroid twice is done.

Reply:
Got it. You are right about distance.